File: C++/Searching/BinarySearch.cpp
class BinarySearch
{
public:
    static int binarySearchIterative(int arr[], int size, int num)
    {
        return binarySearchIter(arr, num, 0, size - 1);
    }

private:
    static int binarySearchIter(int arr[], int num, int begin, int end)
    {
        while (begin <= end)
        {
            int mid = (begin + end) / 2;
            if (arr[mid] < num)
                begin = mid + 1;
            else if (arr[mid] > num)
                end = mid - 1;
            else
                return mid;
        }
        return -1;
    }

public:
    static int binarySearchRecursive(int arr[], int size, int num)
    {
        return binarySearchRecur(arr, num, 0, size - 1);
    }

private:
    static int binarySearchRecur(int arr[], int num, int begin, int end)
    {
        if (begin > end)
            return -1;
        int mid = (begin + end) / 2;
        if (arr[mid] == num)
            return mid;
        if (arr[mid] > num)
            return binarySearchRecur(arr, num, begin, mid - 1);
        else
            return binarySearchRecur(arr, num, mid + 1, end);
    }
};

File: Rust/src/searching/binary_search.rs
/// Binary search algorithm
/// ### Arguments
/// * `arr` - A vector of items to search
/// * `item` - The item to search for
/// ### Returns
/// * `Option<usize>` - The index of the item if found, otherwise None
/// ### Example
/// ```
/// use rust::searching::binary_search::binary_search;
/// let arr = vec![1, 2, 3, 4, 5];
/// let item = 3;
/// let index = binary_search(arr, item);
/// assert_eq!(index, Some(2));
/// ```
/// ### Time Complexity
/// `O(log n)`
/// ### Space Complexity
/// `O(1)`
pub fn binary_search<T: PartialOrd + Copy>(arr: Vec<T>, item: T) -> Option<usize> {
    let (mut low, mut high) = (0, arr.len() - 1);

    while low <= high {
        let mid = (low + high) / 2;

        if arr[mid] == item {
            return Some(mid);
        } else if arr[mid] > item {
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }

    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_binary_search() {
        assert_eq!(binary_search(vec![1, 2, 3, 4, 5], 3), Some(2));
        assert_eq!(binary_search(vec!['a', 'b', 'c', 'd', 'e'], 'c'), Some(2));
    }

    #[test]
    fn test_binary_search_not_found() {
        assert_eq!(binary_search(vec![1, 2, 3, 4, 5], 6), None);
    }
}
